rfcs/0009-decompose-ledger-builder.md

```diff
@@ -0,0 +1,66 @@
+## Summary
+[summary]: #summary
+
+This RFC proposes a decoupling of the ledger builder which is will make the components of the ledger builder more composable and pave the way for properly encoding the ledger builder into the new transition frontier data structure.
+
+## Motivation
+[motivation]: #motivation
+
+The ledger builder has been suffering from scope creep for some time now. As we are moving towards the new transition frontier data structure, we are already decoupling the storage mechanism for the underlying ledger for a ledger builder. This is going to have a number of difficult ramifications on the existing data structure, so now seems as good a time as any to fully decouple the ledger builder.
+
+## Detailed design
+[detailed-design]: #detailed-design
+
+[Full architecture](../docs/res/all_data_structures.png)
+
+![](../docs/res/ledger_builder_data_structures.png)
+
+### `Merkle_ledger`
+
+A `Merkle_ledger.t` is a value and a first class module implementing a `Merkle_ledger_intf`, which is a common subset of all ledgers. Having a first class module representation of a ledger allows us to build code which can be generic over an interface without functoring over it.
+
+### `Merkle_ledger_diff`
+
+A `Merkle_ledger_diff.t` is a compressed difference from a base ledger to a target ledger. It can be applied to a base ledger to make it a target ledger. More formally, a `Merkle_ledger_diff.t` is a base ledger, a target ledger, and a set of updated accounts, and a set of new accounts. The set of updated accounts represents all of the accounts that have changed between the base and the target, while the set of new accounts contains new accounts to add along with the locations in the tree to add them to. A `Merkle_ledger_diff.t` can be applied to a ledger by iterating through each set of accounts and writing them to the ledger. After applying to a ledger, the root of the ledger should be equal to the target merkle root iff the root before applying was equal to the base merkle root.
```

Comment:
In the future, would `Merkle_ledger_diff.t` include deleted accounts?

Reply:
Yes. `Merkle_ledger_diff.t` should have the ability to act as a transition between any two possible states of a `Merkle_ledger`. In this sense, I guess you could say it's intended to be universal, so whenever we add a new feature like account deletion, this would need to get updated as well.